### ISSUE TITLE
Rewrite `arithmetic_side_effects` to not use matching on string representation of types

### ISSUE THREAD

**Issue by samueltardieu** [rust-lang/rust-clippy#16359]
### Description

`arithmetic_side_effects` uses the string representation of types to compare them. This will miss aliases (including `core`/`alloc` → `std`), takes longer to compare, etc.

It should use a method similar to what is done in the `disallowed_types` lint, where maps of `DefId` or `PrimTy` are built at lint initialization time and then used throughout the tree walk. Some type lookup code could probably even be shared (including the part where unloaded crates don't cause errors).

### Version

```text

```

**Comment by keirsalterego** [comment]
@rustbot claim

I'd like to work on it!
